debian/patches/0004-add-cuda-tonemap-impl.patch

```diff
@@ -276,6 +276,33 @@ Index: FFmpeg/libavfilter/colorspace.h
 +#define ARIB_B67_C 0.55991073f
 +#define FLOAT_EPS 1e-6f
 +
++/*
++ * Pre-calculated constants used for YCbCr narrow to full range scaling
++ * The base formula is the quantization formula derived from BT.2100 Table 9:
++ * Where Y' = Round [(219 * E′ + 16) * 2^(n−8)],
++ * Cb',Cr' = Round [(224 * E′ + 128) * 2^(n−8)]
++ * where E' is the signal value in [0,1] range and n is the bit depth. Round is rounding towards 0.
++ * For inputs, the inverse is used where we are solving for E' for a given Y'Cb'Cr' normalized by GPU
++ * in [0,1] range. The GPU will interpret color as a 16bit int value, and solving for E' becomes:
++ * E' = (Y' - 2^(n-4)) / (219 * 2^(n-8))
++ * E' = (Cb'Cr' - 2^(n-1)) / (7 * 2^(n-3))
++ * Y' and Cb'Cr' is in the range of [0, 2^n - 1] in original formula, we need to scale the value normalized to [0,1]:
++ * C = Y'Cb'Cr' * (2^n - 1)
++ * Which means the input scale = (2^n - 1) / (219 * 2^(n-8)) and input offset = 2^(n-4)) / (219 * 2^(n-8)) for Y' and
++ * 2^(n-1)) / (7 * 2^(n-3)) for Cb'Cr'
++ */
++#define INPUT_Y_SCALE_10B 1.1678082192f
++#define INPUT_UV_SCALE_10B 1.1417410714f
++#define INPUT_Y_SCALE_12B 1.1689497717f
++#define INPUT_UV_SCALE_12B 1.1428571429f
++
++/*
++ * GPU will interpret 10bit and 12bit color as 16bit int
++ * but that will introduce a slight (2^(16-n))/2^16 quantization offset which we want to compensate for
++*/
++#define QUANTIZATION_OFFSET_10B 0.0009765774014f
++#define QUANTIZATION_OFFSET_12B 0.0002441443503f
++
```

Comment:
Since we are going to put them in a public area, why not use macro? It also avoids calculation errors.
```c
#include <stdio.h>

#define INPUT_Y_SCALE(n)  ((double)((1 << (n)) - 1) / (219 * (1 << ((n) - 8))))
#define INPUT_UV_SCALE(n) ((double)((1 << (n)) - 1) / (224 * (1 << ((n) - 8))))
#define QUANTIZATION_OFFSET(n) ((double)(1 << (16 - (n))) / ((1 << 16) - 1))

int main()
{
   printf("%.12lff\n", INPUT_Y_SCALE(10));
   printf("%.12lff\n", INPUT_UV_SCALE(10));
   printf("%.12lff\n\n", QUANTIZATION_OFFSET(10));

   printf("%.12lff\n", INPUT_Y_SCALE(12));
   printf("%.12lff\n", INPUT_UV_SCALE(12));
   printf("%.12lff\n\n", QUANTIZATION_OFFSET(12));
   return 0;
}
```
```
1.167808219178f
1.141741071429f
0.000976577401f

1.168664383562f
1.142578125000f
0.000244144350f
```

Also, there does exist `yuv420p12le` which can be mapped to `P012`. This is a format used by intel RExt.

Reply:
> Also, there does exist yuv420p12le which can be mapped to P012

Our filters do not handle such frames atm though

Reply:
`sed -i 's/AV_PIX_FMT_P016/AV_PIX_FMT_P012/g' libavcodec/cuviddec.c libavcodec/nvdec.c libavfilter/*cuda*`

In this way we can ditch the misuse of P016. P012 was added only in recent years.

Reply:
That's a too aggressive change though, I'd prefer to leave it for future.